WEB_SERVICE/WEB/_MEDIUM/GRAPHQL_CIRCULAR_FRAGMENTS/recommendation.md

```diff
@@ -0,0 +1,132 @@
+npx graphql-codegen init
+```
+
+You can configure the code generator to output schema validations that detect potential circular fragment issues in your GraphQL queries.
+
+### 2. Implement Query Cost Analysis
+Use query complexity analysis to assign a "cost" to each field, fragment, or query. When a query or fragment exceeds a predefined cost threshold, it can be automatically rejected.
+
+ - **Tooling:** You can use libraries like graphql-query-complexity to calculate query cost and limit the impact of circular fragments.
+
+**Example in JavaScript using graphql-query-complexity:**
+
+```javascript
+const { getComplexity, simpleEstimator } = require('graphql-query-complexity');
+
+const server = new ApolloServer({
+  schema,
+  validationRules: [
+    queryComplexity({
+      estimators: [
+        simpleEstimator({ defaultComplexity: 1 })
+      ],
+      maximumComplexity: 1000,  // Queries that exceed this complexity will be rejected
+      onComplete: (complexity) => {
+        console.log('Query Complexity:', complexity);
+      },
+    }),
+  ],
+});
+```
+
+### 3. Enforce Depth and Recursion Limits
+
+Some GraphQL server implementations allow you to define limits for query depth and recursion to prevent resource exhaustion. This helps avoid infinite loops caused by circular fragments.
+
+ - **Tooling:** Apollo Server, for example, supports query depth limits, which can prevent excessive recursion.
+
+**Example using query depth limit in Apollo Server:**
+
+```javascript
+const { ApolloServer } = require('apollo-server');
+const depthLimit = require('graphql-depth-limit');
+
+const server = new ApolloServer({
+  schema,
+  validationRules: [depthLimit(5)],  // Set query depth limit to 5
+});
+```
+
+### 4. Use Fragments Wisely
+
+Avoid overusing fragments in complex queries. Overuse can lead to performance degradation and opens up possibilities for circular references. Instead, consider splitting complex queries into multiple smaller queries.
+
+ - **Recommendation:** Review query design and limit fragment usage to specific, modular pieces of data.
+
+**Example:**
+
+```graphql
+query {
+  user(id: "1") {
+    name
+    email
+    posts {
+      title
+      content
+    }
+  }
+}
+```
+
+Rather than nesting multiple fragments within each other, use a simpler query like this.
+
+### 5. Rate Limiting to Prevent Exploitation
+
+Apply rate limiting to protect your GraphQL server from being overwhelmed by malicious actors exploiting circular fragments to trigger denial-of-service attacks.
+
+ - **Tooling:** Use rate-limiting middleware, such as express-rate-limit, to restrict the number of requests from a single source over a certain period.
+
+**Example using** `express-rate-limit`:
+
+```javascript
+const rateLimit = require('express-rate-limit');
+
+const limiter = rateLimit({
+  windowMs: 15 * 60 * 1000, // 15 minutes
+  max: 100,  // limit each IP to 100 requests per windowMs
+});
+
+app.use('/graphql', limiter);
+```
```

Comment:
This will not stop the attack because it needs only one request , so I think it should be removed